### PR TITLE
Feature gcode execution result triage

### PR DESCRIPTION
**Pull request by snapmaker-lixiaolong**

### CODE REVIEW

File: snapmaker/src/common/debug.cpp [hunk 1]
```diff
@@ -198,17 +198,27 @@ void SnapDebug::CmdChecksumError(bool screen) {
...
+  // Log(SNAP_DEBUG_LEVEL_INFO, "Last recv line: %d\n", systemservice.current_line());
+  // Log(SNAP_DEBUG_LEVEL_INFO, "Last ack line: %d\n", info.last_line_num_of_sc_gcode);
+  // Log(SNAP_DEBUG_LEVEL_INFO, "Last st line: %d\n", pl_recovery.LastLine());
+  // Log(SNAP_DEBUG_LEVEL_INFO, "Fault: 0x%08X, action ban: 0x%X, power ban: 0x%X\n",
+  //       systemservice.GetFaultFlag(), action_ban, power_ban);
+  // Log(SNAP_DEBUG_LEVEL_INFO, "Homing: 0x%X, axes_known: 0x%X\n", axis_homed, axis_known_position);
+  // Log(SNAP_DEBUG_LEVEL_INFO, "active coordinate: %d\n", gcode.active_coordinate_system);
+  // Log(SNAP_DEBUG_LEVEL_INFO, "coordinate 1: X: %.3f, Y: %.3f, Z: %.3f, B: %.3f\n",
+  //     gcode.coordinate_system[0][X_AXIS], gcode.coordinate_system[0][Y_AXIS], gcode.coordinate_system[0][Z_AXIS], gcode.coordinate_system[0][B_AXIS]);
+
```
Comment (xhycode) [inline snapmaker/src/common/debug.cpp:212]: Useless code should be deleted

File: Marlin/src/gcode/queue.cpp [hunk 2]
```diff
@@ -249,7 +249,9 @@ void ok_to_send() {
   if(Screen_send_ok[cmd_queue_index_r]) {
     ack_gcode_event(Screen_send_ok_opcode[cmd_queue_index_r], CommandLine[cmd_queue_index_r]);
     SNAP_DEBUG_SET_GCODE_LINE(CommandLine[cmd_queue_index_r]);
-    Screen_send_ok[cmd_queue_index_r] = false;
+    // Screen_send_ok[cmd_queue_index_r] = false;
```
Comment (scotthsl) [inline Marlin/src/gcode/queue.cpp:252]: why comment this line？

File: snapmaker/src/gcode/M1005_1007.cpp [hunk 3]
```diff
@@ -44,12 +44,22 @@ void GcodeSuite::M1005() {
   LOG_I("%s %s\n", MSG_MARLIN, SHORT_BUILD_VERSION);
   LOG_I("Compiled: %s, %s\n", __DATE__, __TIME__);
 
+  // SERIAL_ECHOLN(MSG_MARLIN);
+  // SERIAL_ECHOLN(SHORT_BUILD_VERSION);
+  // SERIAL_ECHOLN("\n");
```
Comment (scotthsl) [inline snapmaker/src/gcode/M1005_1007.cpp:49]: no need to reserve these comments

File: snapmaker/src/gcode/M1005_1007.cpp [hunk 4]
```diff
@@ -44,12 +44,22 @@ void GcodeSuite::M1005() {
...
+  SERIAL_ECHOPAIR(MSG_MARLIN, SHORT_BUILD_VERSION, "\n");
+  SERIAL_ECHOPAIR("Compiled: ", __DATE__, ", ", __TIME__, "\n");
+
   // version in package
   memcpy(buffer, (char*)(FLASH_BOOT_PARA + 2048), 30);
-  LOG_I("%s: %s\n", MSG_MARLIN_PACK, buffer);
+  // LOG_I("%s: %s\n", MSG_MARLIN_PACK, buffer);
```
Comment (scotthsl) [inline snapmaker/src/gcode/M1005_1007.cpp:55]: no need to reserve this comment

File: snapmaker/src/gcode/M1005_1007.cpp [hunk 5]
```diff
@@ -44,12 +44,22 @@ void GcodeSuite::M1005() {
...
+
+  SERIAL_ECHOPAIR(MSG_MARLIN_PACK, buffer, "\n");
+
 
   // version of modules
-  LOG_I("Module Ver:\n");
+  // LOG_I("Module Ver:\n");
```
Comment (scotthsl) [inline snapmaker/src/gcode/M1005_1007.cpp:61]: no need to reserve this comment

File: snapmaker/src/hmi/gcode_result_handler.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+      gcode_result_mgt_.index = 0;
+    }
+
+    void GcodeResultCheck(unsigned char ch);
+    void GcodeResultFlush();
+  private:
+    screen_gcode_result_mgt_t gcode_result_mgt_;
+};
+
+extern GcodeResultHandler gcode_result_handler;
```
Comment (scotthsl) [inline snapmaker/src/hmi/gcode_result_handler.h:24]: here should adda  new line

File: snapmaker/src/gcode/M1005_1007.cpp (hunk 4 above)
Comment (747lulu747) [reply snapmaker/src/gcode/M1005_1007.cpp:55]: Why use "SERIAL_ECHOPAIR(MSG_MARLIN_PACK, buffer, "\n")" instand of "LOG_I("%s: %s\n", MSG_MARLIN_PACK, buffer);"?
What is the difference between LOG_I and SERIAL_ECHOPAIR?

**Comment by 747lulu747** [comment]
It work OK but not the best solution. Here is the disadvantage:
1) HardwareSerial is a base standard module. It should NOT been changed. Other developer may make  a mistake when using this Change-HardwareSerial module.

Here is a suggestion:
Redefine the follow
`#define SERIAL_OUT(WHAT, ...) (void)MYSERIAL0.WHAT(__VA_ARGS__)`
Change it to
```
#define SERIAL_OUT(WHAT, ...)                 {\   
                                                                      (void)MYSERIAL0.WHAT(__VA_ARGS__);\
                                                                      (void)GCODE.WHAT(__VA_ARGS__);\
                                                                  } while(0)
```

All the SERRIAL_OUT will call the GCODE_OUT. And following API to "GcodeResultHandler.cpp" :
```
write(char x)
print(x)
print(__VA_ARGS__)
println(x)
print(x, b)
println(x, b)
printf(__VA_ARGS__)
flush()
```

File: Marlin/src/gcode/queue.cpp (hunk 2 above)
Comment (747lulu747) [reply Marlin/src/gcode/queue.cpp:252]: Screen_send_ok[cmd_queue_index_r] will set to false when the data in sending buffer have send out  actually.
Here keep Screen_send_ok[cmd_queue_index_r] as true means that there is data in buffer need to sendout.

**Comment by scotthsl** [comment]
PR improved by xiehongyan has been merged, so close this.

**Pull request closed without merging**